### PR TITLE
Recognition for trivial I-bundle over torus

### PR DESCRIPTION
**Pull request by bobbycyiii**
I've implemented a test for homeomorphism to T^2 x I, and made unit tests and Python bindings for it, too.

### CODE REVIEW

File: engine/triangulation/dim3/decompose.cpp [hunk 1]
```diff
@@ -561,7 +564,7 @@ bool Triangulation<3>::isSolidTorus() const {
         delete crushed;
     }
 }
-
+
```
Comment (WPettersson) [inline engine/triangulation/dim3/decompose.cpp:567]: Extra spacing accidentally added

**Comment by WPettersson** [comment]
Looks good to me, other than that odd space added.

**Comment by baburton** [comment]
Hi - I was looking to merge this into master, but I have a couple of concerns. I can try to fix them if I have time, but if not, they are:

- Less important: I don't think you need a zero-efficient triangulation before you do your fillings, and *making* a zero-efficient triangulation (even if it was zero-efficient to begin with) is expensive because of the normal surface enumeration. I'd suggest just using intelligentSimplify() and then make the boundaries one-vertex as you already do, and then go straight to the fillings. The subsequent calls to isSolidTorus() will also try to avoid normal surface enumeration if possible, and if not, hopefully the additional filling has made the triangulation smaller so the normal surface enumeration becomes cheaper.

- More important: I don't think the one-vertex-boundary code is guaranteed to succeed. I believe I can prove it works on all cases except for when the embedded boundary edge hits a vertex with boundary degree one; in that case you cannot layer, and the call to working->layerOn(emb) will result in undefined behaviour due to the failed precondition that the triangles on either side of the edge are distinct. Note that it *is* possible to create a two-vertex torus where the only embedded edge touches a degree one vertex.

This second problem is solvable: if you have a degree one boundary vertex then you can layer on the edge *opposite* that vertex, and that will convert your problematic edge into one that you are allowed to layer on, and then your code can continue as normal.

- b.

**Comment by bobbycyiii** [comment]
Degree-one vertices do make the boundary simplification fail. I will correct this error immediately.

The first suggestion looks good to me as well, so I'll do that too.

**Comment by baburton** [comment]
Actually, in my suggested fix, the extra layering is not needed. If you have a degree one boundary vertex then you can immediately do the close-book move on the edge opposite. - b.

**Comment by baburton** [comment]
Hi - I've merged this now into master. Some comments (none of which are serious; just more FYI for the next patch):

- I ended up moving the close-book/layering process into a new routine minimiseBoundary() since this is more generally useful; also I reorganised the way it works so that it priorities close-book moves over everything else (since this means not increasing the number of tetrahedra where possible).

- I claim in the documentation for mimimiseBoundary() that it will never break 0-efficiency. I believe I've proven this, but the proof was all in my head and so I'm happy if you want to independently verify that you agree with this. I added this because it's a useful promise to make for some other higher-level routines that genuinely require 0-efficiency.

- I've made knowsTorusXInterval() a bit more functional, in that it now performs several trivial tests (validity, connectedness, etc.).

- The documentation for the member variable torusXInterval_ was wrong (please be careful with copy-and-paste!).

- You don't need headers triangulation/generic/*, since dim3.h will pull all that in. You also don't need homologicaldata.h, since that's giving more heavy-duty algebraic code that you're not using (the basically homology functions are already in the triangulation header).

- I renamed the functions isTxI and knowsTxI, after many bad experiences with choosing long mixed-case function names that were a pain to type and difficult to change after several years. :)

- In the docs for isTxI(), I ended the sentence before the journal reference since doxygen pulls out the first sentence as a quick summary when generating API docs. I also removed the arXiv reference since JoCG is free for all readers.

- I removed the one-vertex precondition, since minimiseBoundary() guarantees this.

- In the homology tests, it's generally not good to test using string comparison (since this assumes the pretty-printed string form of a group will never change). I've changed this to test rank and torsion factors directly.

- I removed the homologyBdry() test, since we already check that the boundary is two disjoint tori, and so there's nothing extra to gain here.

Finally: I'm not super comfortable with deliberately using closeBook() in a way that violates its preconditions (to do the fillings). I changed this to a manual join/unjoin, which will also have the side-effect of not cloning the triangulation three times over.

Anyway, that's all! It's a long list, but mainly I want to let you know the kinds of things I'm doing here before merging code.

And finally, please do feel very free to take a look at the code in its current form (see git master) to see if I broke anything. :)

- b.

**Comment by baburton** [comment]
Closing this now since version 6.0 (being released in the next 24 hours) includes TxI recognition. Not merging this branch directly because (i) this branch derives from 6.0_pre and not master, and (ii) the TxI recognition in 6.0 is a hybrid of cherry-picked code from here and independently written code.

**Pull request closed without merging**